rust/src/lib.rs

```diff
@@ -170,17 +234,216 @@ pub unsafe extern "C-unwind" fn create_client(
+            (None, &strings[0], &strings[1], 8u32)
+        }
+        (redis::PushKind::Disconnection, _) => {
+            logger_core::log(
+                logger_core::Level::Info,
+                "pubsub",
+                "PubSub disconnection received",
+            );
+            return;
+        }
+        (kind, len) => {
+            logger_core::log(
+                logger_core::Level::Error,
+                "pubsub",
+                &format!(
+                    "Invalid PubSub message structure: kind={:?}, len={}",
+                    kind, len
+                ),
+            );
+            return;
+        }
+    };
+
+    // Prepare pointers while keeping strings alive
+    let pattern_ptr = pattern.map(|p| p.as_ptr()).unwrap_or(std::ptr::null());
+    let pattern_len = pattern.map(|p| p.len() as i64).unwrap_or(0);
+    let channel_ptr = channel.as_ptr();
+    let channel_len = channel.len() as i64;
+    let message_ptr = message.as_ptr();
+    let message_len = message.len() as i64;
+
+    // Call callback while strings are still alive
+    unsafe {
+        pubsub_callback(
+            kind,
+            message_ptr,
+            message_len,
+            channel_ptr,
+            channel_len,
+            pattern_ptr,
+            pattern_len,
+        );
+    }
+
+    // Vec<u8> instances are automatically cleaned up here
+    // No memory leak, no use-after-free
+}
+
 /// Closes the given client, deallocating it from the heap.
 /// This function should only be called once per pointer created by [`create_client`].
 /// After calling this function the `client_ptr` is not in a valid state.
 ///
+/// Implements graceful shutdown coordination for PubSub tasks with timeout.
+///
 /// # Safety
 ///
 /// * `client_ptr` must not be `null`.
 /// * `client_ptr` must be able to be safely casted to a valid [`Arc<Client>`] via [`Arc::from_raw`]. See the safety documentation of [`Arc::from_raw`].
 #[unsafe(no_mangle)]
 pub extern "C" fn close_client(client_ptr: *const c_void) {
     assert!(!client_ptr.is_null());
+
+    // Get a reference to the client to access shutdown coordination
+    let client = unsafe { &*(client_ptr as *const Client) };
+
+    // Take ownership of shutdown sender and signal graceful shutdown
+    if let Ok(mut guard) = client.pubsub_shutdown.lock() {
+        if let Some(shutdown_tx) = guard.take() {
+            logger_core::log(
+                logger_core::Level::Debug,
+                "pubsub",
+                "Signaling PubSub task to shutdown",
+            );
+
+            // Send shutdown signal (ignore error if receiver already dropped)
+            let _ = shutdown_tx.send(());
+        }
+    }
+
+    // Take ownership of task handle and wait for completion with timeout
+    if let Ok(mut guard) = client.pubsub_task.lock() {
+        if let Some(task_handle) = guard.take() {
+            let timeout = std::time::Duration::from_secs(5);
```

Comment:
Should this be passed down from `PubSubPerformanceConfig.ShutdownTimeout instead of hard-coded`?